### PR TITLE
Implement NA.__array_ufunc__

### PR DESCRIPTION
**Pull request by TomAugspurger**
This gives us consistent comparisons with NumPy scalars.

I have a few implementation questions:

1. I wanted to reuse `maybe_dispatch_ufunc_to_dunder_op`. That was in `core/ops/dispatch.py`, but NA is defend in `_libs/missing.pyx`. For now I just moved it to `_libs/missing.pyx`, is there a better home for it? I don't think `_libs/ops.pyx` is an option, as it imports missing.
2. How should we handle comparisons with ndarrays? I see a few options:

  a. Return an `ndarray[object]`, where each element is NA
  b. Really take over and return our extension arrays. So `np.array([1, 2]) == pd.NA` would return an `IntegerArray([NA, NA])`. Boolean results would return a BooleanArray. And float results would return a ... PandasArray? (so maybe this isn't a good idea. But worth considering)
3. What scalars should we handle (in `__array_ufunc__` and in the regular methods)? Any of date time, Timestamp, time delta, Period, Interval? 

One item 2, I seem to have messed something up, as we return just a scalar `NA` right now. Will want to sort that out.

### CODE REVIEW

**Comment by TomAugspurger** [comment]
> How should we handle comparisons with ndarrays?

https://github.com/pandas-dev/pandas/pull/30245/commits/46f2327fa36dd110f5adae311dce487a06dc03f6 restores the behavior on master. We return an object-dtype ndarray filled with NA values. Not sure if that's the right choice though.

**Comment by TomAugspurger** [comment]
Anybody else seeing the CI check failure when linting `.pyx` files?

```
flake8 --version
3.7.9 (flake8-comprehensions: 3.1.4, mccabe: 0.6.1, pycodestyle: 2.5.0, pyflakes: 2.1.1) CPython 3.7.3 on Linux
Linting .py code
Linting .py code DONE
Linting .pyx code
Traceback (most recent call last):
  File "/home/runner/miniconda3/envs/pandas-dev/bin/flake8", line 11, in <module>
    sys.exit(main())

 ...

    self.formatter.handle(error)
  File "/home/runner/miniconda3/envs/pandas-dev/lib/python3.7/site-packages/flake8/formatting/base.py", line 91, in handle
    line = self.format(error)
  File "/home/runner/miniconda3/envs/pandas-dev/lib/python3.7/site-packages/flake8/formatting/default.py", line 39, in format
    "col": error.column_number,
ValueError: unsupported format character ':' (0x3a) at index 41
```

Not able to reproduce locally with those versions.

**Comment by TomAugspurger** [comment]
All green now. Any thoughts on what, e.g. `np.array([True, False]) | pd.NA` should return? Our choices are an object-dtype ndarray, or a BooleanArray. Is it strange to return a non-ndarray?

File: doc/source/user_guide/missing_data.rst
Comment (jreback) [inline doc/source/user_guide/missing_data.rst]: there is a section in dsintro.rst that references ufuncs, I would at the very least reference it.

**Review by jreback — CHANGES_REQUESTED** [review]
will review soon

**Comment by jorisvandenbossche** [comment]
> How should we handle comparisons with ndarrays? I see a few options:

Another option might also be to raise right now? (or is that also surprising?) That might be the safest option to allow choosing a more specific behaviour later. 

Otherwise returning an object array of NAs seems OK to me. But from numpy's point of view, such an object array is also strange, as you can't do anything boolean with it like boolean indexing.

File: doc/source/user_guide/missing_data.rst
Comment (jorisvandenbossche) [inline doc/source/user_guide/missing_data.rst]: Didn't yet check in detail the PR, but does this then also special case for np.power / np.divide ? (similar to what we have been discussing lately for eg IntegerArray)

File: doc/source/user_guide/missing_data.rst
Comment (jorisvandenbossche) [inline doc/source/user_guide/missing_data.rst]:
```suggestion
   Currently, ufuncs involving an ndarray an NA will return an
```
```suggestion
   Currently, ufuncs involving an ndarray and NA will return an
```

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [inline pandas/_libs/missing.pyx]: Why are the `np.int64, np.int_` needed? `numbers.Number` should already deal with those I think, and also adding only int64 seems strange

**Comment by TomAugspurger** [comment]
> But from numpy's point of view, such an object array is also strange, as you can't do anything boolean with it like boolean indexing.

Right. An ndarray full of pd.NA is a fairly useless thing I think. That's my main motivation for considering anything fancy like returning an IntegerArray, etc.

We do have precedence for returning non-ndarrays from `__array_ufunc__`.

```python
In [8]: a = pd.array([2000, 2001], dtype="Period[D]")

In [9]: np.add(a, 1)
Out[9]: array([Period('2000-01-02', 'D'), Period('2001-01-02', 'D')], dtype=object)
```

Raising for ndarrays is also fine by me. Just need to choose whether a 0-dim ndarray is an array or scalar (probably a scalar).

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [inline pandas/_libs/missing.pyx]: This seems a bit strange, but I suppose it is to follow numpy behaviour of 0-dim arrays returning scalars from comparison operations? (if so, maybe add a comment about that)

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [inline pandas/_libs/missing.pyx]: Timestamp et al don't implement `__array_ufunc__` right? If they would, I would say they could handle NA instead of NA handling all those types.. 

(I also think it is not that important right now that pd.NA can handle those types, since we don't yet use NA for those dtypes)

**Review by jorisvandenbossche — COMMENTED** [review]
(only looked at the first part)

**Comment by jorisvandenbossche** [comment]
> We do have precedence for returning non-ndarrays from __array_ufunc__.

What you show is actually returning a numpy object array (but I would agree that should/could return a PeriodArray)

**Comment by TomAugspurger** [comment]
Oh... i'ts too early here :) I could have sworn we returned an EA there.

File: doc/source/user_guide/missing_data.rst
Comment (TomAugspurger) [reply doc/source/user_guide/missing_data.rst]: Yes. They dispatch to the dunder op

```python
In [20]: np.power(pd.NA, np.array([0, 1]))
Out[20]: array([1, NA], dtype=object)
```

File: pandas/_libs/missing.pyx
Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: > to follow numpy behaviour of 0-dim arrays returning scalars from comparison operations?

It's to NumPy scalars. Without this, we'd have

```python
np.int64(1) == pd.NA
```

raise with

```pytb
>   out[:] = NA
E   IndexError: too many indices for array
```

**Comment by TomAugspurger** [comment]
Thanks for the review @jorisvandenbossche. Things should be good here.

I've deferred on returning anything too exotic from NA comparisons with ndarrays. We just return an ndarray of NA now.

**Comment by jbrockmendel** [comment]
should getting this in for 1.0 be a priority?

**Comment by jorisvandenbossche** [comment]
> should getting this in for 1.0 be a priority?

All open NA-related PRs are blockers for 1.0 I think

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: A numpy scalar (which is different from a 0-dim array) is not a ndarray, so wouldn't pass the `isinstance(other, np.ndarray)` test. So I don't fully understand how that example is related.

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [inline pandas/_libs/missing.pyx]: When does it happen that the dispatch function returns NotImplemented?

File: pandas/_libs/missing.pyx
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx]: util.is_bool_object for the second one of these.  Is numbers.Number something like Union[float, int, complex] that we could use the optimizied funcs for?

util.is_array for the isinstance(other, np.ndarray) check.  Even though it is syntactically OK without extra parens, can you put a pair around the isinstance+shape check

File: pandas/_libs/missing.pyx
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx]: can `divmod` be changed to `is_divmod`?  dont want to overlap with built-in names

File: pandas/tests/scalar/test_na_scalar.py
Comment (jbrockmendel) [inline pandas/tests/scalar/test_na_scalar.py]: assuming this isnt fixed before merge, "# FIXME: dont leave commented out" pls

File: pandas/_libs/missing.pyx
Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: Numbers is appropriate since it’s an ABC others can register on.

File: pandas/_libs/missing.pyx
Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: Did I define that here or is it called that on master?

File: pandas/_libs/missing.pyx
Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: The comment just needs to be removed. It’s handled correctly I think. 

It’s not implemented for ufuncs that aren’t dunder ops.

File: pandas/_libs/missing.pyx
Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: its like that on master too, still should be changed.  OK for separate PR if thats a hassle

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: I personally prefer the explicit `isinstance` checks here. It is already complex to fully oversee what types are handled, and using those helper methods only makes it more opaque.

**Comment by TomAugspurger** [comment]
FYI I’ll be offline most of the next week. Feel free to push commits here, otherwise I’ll pick this up when I can.

File: pandas/_libs/missing.pyx
Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: in general i agree, but in the cython code we generally lean towards optimizing harder.  no strong opinion.

Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: That's also a good point (cython does not optimize those automatically like that I suppose?). But note that this is not the most performance sensitive code (it's not in the inner loop of a vectorized operation or so)

**Comment by TomAugspurger** [comment]
This should be good to go, assuming CI passes.

**Comment by TomAugspurger** [comment]
All green.

**Comment by TomAugspurger** [comment]
Planning to merge this later today if there aren't any objections.

**Review by jreback — COMMENTED** [review]
lgtm, except for a question.

File: pandas/_libs/ops_dispatch.pyx
Comment (TomAugspurger) [reply pandas/_libs/ops_dispatch.pyx]: There’s not a performance reason. The only reason is to avoid a Python import inside the cython function. Do you have a preference?

File: pandas/_libs/missing.pyx
Comment (jbrockmendel) [reply pandas/_libs/missing.pyx]: could call lib.item_from_zerodim at the top?

File: pandas/_libs/ops_dispatch.pyx
Comment (jbrockmendel) [inline pandas/_libs/ops_dispatch.pyx]: fstr here and on 87

File: pandas/_libs/ops_dispatch.pyx
Comment (jbrockmendel) [inline pandas/_libs/ops_dispatch.pyx]: for functions that arent `cdef` or `cpdef`, we can use python syntax for annotations

File: pandas/_libs/ops_dispatch.pyx
Comment (jbrockmendel) [reply pandas/_libs/ops_dispatch.pyx]: Agreed avoiding the import is worthwhile

Comment (jreback) [reply pandas/_libs/ops_dispatch.pyx]: ok fair.

**Review by jreback — APPROVED** [review]
lgtm. @jbrockmendel

File: pandas/_libs/missing.pyx
Comment (jbrockmendel) [inline pandas/_libs/missing.pyx]: looks like we should have 482 or 483 but not both?

**Comment by jbrockmendel** [comment]
comment about L482 vs L483, otherwise LGTM

File: pandas/_libs/missing.pyx
Comment (TomAugspurger) [reply pandas/_libs/missing.pyx]: With this

```diff
diff --git a/pandas/_libs/missing.pyx b/pandas/_libs/missing.pyx
index 8d4d2c5568..e2bb6448cd 100644
--- a/pandas/_libs/missing.pyx
+++ b/pandas/_libs/missing.pyx
@@ -295,8 +295,7 @@ def _create_binary_propagating_op(name, is_divmod=False):
 
     def method(self, other):
         if (other is C_NA or isinstance(other, str)
-                or isinstance(other, (numbers.Number, np.bool_))
-                or isinstance(other, np.ndarray) and not other.shape):
+                or isinstance(other, (numbers.Number, np.bool_))):
             # Need the other.shape clause to handle NumPy scalars,
             # since we do a setitem on `out` below, which
             # won't work for NumPy scalars.
```

I have

```pytb
In [3]: np.int64(1) == pd.NA
/Users/taugspurger/.virtualenvs/pandas-dev/bin/ipython:1: DeprecationWarning: elementwise comparison failed; this will raise an error in the future.
  #!/Users/taugspurger/Envs/pandas-dev/bin/python
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-3-209227154583> in <module>
----> 1 np.int64(1) == pd.NA

~/sandbox/pandas/pandas/_libs/missing.pyx in pandas._libs.missing._create_binary_propagating_op.method()
    307         elif isinstance(other, np.ndarray):
    308             out = np.empty(other.shape, dtype=object)
--> 309             out[:] = NA
    310
    311             if is_divmod:

IndexError: too many indices for array
```

So when we get there, it really does seem like `np.int64(1)` is an ndarray. Is cython doing something?

**Comment by TomAugspurger** [comment]
Fixed. Merging once https://github.com/pandas-dev/pandas/pull/30245/files/0f4e121a229e3ede7d3b0fa4a974f5e2dcc8e30e#diff-c4b0cf199cc6d5bc2362ceb737159458 is resolved @jorisvandenbossche.

**Comment by TomAugspurger** [comment]
Merging in a few hours if there aren't any objections. Can revisit the line in https://github.com/pandas-dev/pandas/pull/30245/files/0f4e121a229e3ede7d3b0fa4a974f5e2dcc8e30e#diff-c4b0cf199cc6d5bc2362ceb737159458 if needed.

File: pandas/_libs/missing.pyx
Comment (jorisvandenbossche) [reply pandas/_libs/missing.pyx]: > So when we get there, it really does seem like np.int64(1) is an ndarray. Is cython doing something?

Possibly, no idea. But I trust you this is indeed needed to get it working

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Checks are failing: the benchmark run failure seems unrelated, the typing failure looks unrelated as well but I am not very knowledgeable about that.

**Comment by jreback** [comment]
thanks @TomAugspurger very nice!

**Pull request merged**
